plugins/tiddlywiki/sqlite3store/sql-functions.js

```diff
@@ -0,0 +1,217 @@
+/*\
+title: $:/plugins/tiddlywiki/sqlite3store/sql-functions.js
+type: application/javascript
+
+Functions to perform basic tiddler operations with a sqlite3 database
+
+This file is spliced into the HTML file to be executed before the boot kernel has been loaded.
+
+\*/
+
+(function() {
+
+$tw.SqlFunctions = function(options) {
+	options = options || {};
+	var self = this;
+	this.db = new $tw.sqlite3.oo1.DB("/tiddlywiki.sqlite3","c");
+	/*
+	Create tables and indexes
+	*/
+	self.db.exec({
+		sql: `
+		DROP TABLE IF EXISTS tiddlers;
+		CREATE TABLE tiddlers (
+			title TEXT NOT NULL,
+			shadow INTEGER NOT NULL CHECK (shadow = 0 OR shadow = 1), -- 0=real tiddler, 1=shadow tiddler
+			shadowsource TEXT,
+			meta TEXT NOT NULL,
+			text TEXT NOT NULL,
+			PRIMARY KEY(title,shadow)
+		);
+		CREATE INDEX tiddlers_title_index ON tiddlers(title);
+		`
+	});
+	/*
+	Save a tiddler. shadowSource should be falsy for ordinary tiddlers, or the source plugin title for shadow tiddlers
+	*/
+	var statementSaveTiddler = self.db.prepare("replace into tiddlers(title,shadow,shadowsource,meta,text) values ($title,$shadow,$shadowsource,$meta,$text);");
```

Comment:
https://sqlite.org/wasm/doc/trunk/api-oo1.md#stmt
> It is important that statements be finalized in a timely manner, else clients risk introducing locking errors later on in their apps.

I don't have experience with long-lived sqlite connections, so I don't know the behavior from experience, but the above stood out to me.

Reply:
My first implementation used db.exec() but it was super slow - more than a minute to boot the prerelease in Chrome on my machine. Switching to bind got performance to where it is now; very slow but bearable for working with it. But I'm sure there will be better approaches, I'm still feeling my way with sqlite.